presto-main/src/main/java/com/facebook/presto/execution/buffer/LifespanSerializedPageTracker.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.buffer;
+
+import com.facebook.presto.execution.Lifespan;
+import com.facebook.presto.execution.buffer.SerializedPageReference.PagesReleasedListener;
+
+import javax.annotation.Nullable;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+final class LifespanSerializedPageTracker
+        implements PagesReleasedListener
+{
+    private final OutputBufferMemoryManager memoryManager;
+    @Nullable
+    private final PagesReleasedListener childListener;
+    private final ConcurrentMap<Lifespan, AtomicLong> outstandingPageCountPerLifespan = new ConcurrentHashMap<>();
+    private final Set<Lifespan> noMorePagesForLifespan = ConcurrentHashMap.newKeySet();
+    private volatile Consumer<Lifespan> lifespanCompletionCallback;
+
+    public LifespanSerializedPageTracker(OutputBufferMemoryManager memoryManager)
+    {
+        this(memoryManager, Optional.empty());
+    }
+
+    public LifespanSerializedPageTracker(OutputBufferMemoryManager memoryManager, Optional<PagesReleasedListener> childListener)
+    {
+        this.memoryManager = requireNonNull(memoryManager, "memoryManager is null");
+        this.childListener = requireNonNull(childListener, "childListener is null").orElse(null);
+    }
+
+    public boolean isLifespanCompletionCallbackRegistered()
+    {
+        return lifespanCompletionCallback != null;
+    }
+
+    public void registerLifespanCompletionCallback(Consumer<Lifespan> callback)
+    {
+        checkState(lifespanCompletionCallback == null, "lifespanCompletionCallback is already set");
+        this.lifespanCompletionCallback = requireNonNull(callback, "callback is null");
+    }
+
+    public void incrementLifespanPageCount(Lifespan lifespan, int pagesAdded)
+    {
+        AtomicLong counter = outstandingPageCountPerLifespan.get(lifespan);
```

Comment:
isn't `computeIfAbsent` will return the counter in the map if exist? -- so we don't need this extra get and can directly do
```java
counter = outstandingPageCountPerLifespan.computeIfAbsent(lifespan, ignore -> new AtomicLong())
```

Reply:
In JDK8 `ConcurrentHashMap#computeIfAbsent` unconditionally acquires a write lock even when the value is present (and even so, would unconditionally create lambda in any JDK version). Later JDK's address that, but since prestodb still supports JDK8 I chose to go this route.

Reply:
@pettyjamesm : I see. In that case add a comment explain why not directly using `computeIfAbsent` patern. Also since line 67 is not the common path, shall we just use `putIfAbsent` on line 67 instead?

Reply:
`putIfAbsent` unfortunately returns the _previous_ value which makes it more awkward to use in this context than `computeIfAbsent`. I've added a comment about the common case avoiding a lambda creation but omitted the minutia about `ConcurrentHashMap` locking behaviors in JDK 8 since people won't care about that once it's no longer supported.

Reply:
@pettyjamesm : The major motivation of the comment is so maintainers of the code can refactor the code later by reading the comment, what about some comment like following:

```
/  In JDK8 ConcurrentHashMap#computeIfAbsent unconditionally acquires a write lock even when the value is present
// TODO: Remove this get ConcurrentHashMap#get call once Presto no longer supports JDK-8
```